### PR TITLE
improve: commonlib: DiceMatcher SAST (sonar) fixes

### PR DESCRIPTION
**Pull request by kingthorin**
- Reduce cognitive complexity.
- Use .equals() for string comparison.
- Extract assignments from conditionals.
- Declare variables on separate lines.

Signed-off-by: kingthorin <kingthorin@users.noreply.github.com>

### CODE REVIEW

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java [hunk 1]
```diff
@@ -42,43 +42,36 @@ private DiceMatcher() {}
...
      * Author : Jelle Fresen
      * Changes : Fixed indexing to prevent out of array access
+     *           Address various issues identified by static analysis
      * Released under CC-BY-SA.
      */
 
     public static int getMatchPercentage(String a, String b) {
 
+        // Verifying the input:
+        if (a == null || b == null) return 0;
```
Comment (thc202) [inline addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java:52]: Worth adding a test with both null (just to cover all cases).

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java [hunk 2]
```diff
@@ -42,43 +42,36 @@ private DiceMatcher() {}
      * License :  https://creativecommons.org/licenses/by-sa/3.0/
      * Author : Jelle Fresen
      * Changes : Fixed indexing to prevent out of array access
+     *           Address various issues identified by static analysis
      * Released under CC-BY-SA.
      */
 
     public static int getMatchPercentage(String a, String b) {
 
+        // Verifying the input:
```
Comment (thc202) [inline addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java:51]: The comment could be removed.

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java [hunk 3]
```diff
@@ -42,43 +42,36 @@ private DiceMatcher() {}
...
+        if (a == null || b == null) return 0;
+
         String s = a.replaceAll("\\s+", " ");
         String t = b.replaceAll("\\s+", " ");
 
-        // Verifying the input:
-        if (s == null || t == null) return 0;
         // Quick check to catch identical objects:
-        if (s == t) return 1;
+        if (s.equals(t)) return 100;
```
Comment (thc202) [inline addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java:58]: Worth adding a test. The previous check can be kept (which can be tested too).

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java [hunk 4]
```diff
@@ -89,4 +82,18 @@ public static int getMatchPercentage(String a, String b) {
         }
         return (int) Math.floor((double) matches * 100 / (n + m));
     }
+
+    static int[] getBigrams(String str) {
```
Comment (thc202) [inline addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java:86]: `private`

File: addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java [hunk 5]
```diff
@@ -132,4 +136,21 @@ void shouldGiveCorrectPercentageForDifferentString() {
...
+        return Stream.of(
+                Arguments.of((Object) new String[] {null, ORIGINAL_STRING}),
+                Arguments.of((Object) new String[] {ORIGINAL_STRING, null}),
+                Arguments.of((Object) new String[] {"ab", ORIGINAL_STRING}),
+                Arguments.of((Object) new String[] {ORIGINAL_STRING, "ab"}));
+    }
+
+    @ParameterizedTest
+    @MethodSource("stringArraySource")
+    void shouldGiveZeroPercentageForNullOrShortString(String[] in) {
```
Comment (thc202) [inline addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java:150]: - Should be `String stringA, String stringB` (or with other names).
 - `...OrCompletelyDifferentString` does it matter if it's short?

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java (hunk 3 above)
Comment (thc202) [reply addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java:58]: Also, the comment is no longer accurate.

Comment (kingthorin) [reply addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java:58]: You lost me, why would we put the `==` back, that was the SAST issue being addressed.

File: addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java (hunk 5 above)
Comment (kingthorin) [reply addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java:150]: It's an array being passed. Do you mean in the given I should do `String strA = in[a]` then use `strA` for the remainder of the test?

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java (hunk 3 above)
Comment (thc202) [reply addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java:58]: `==` checks identity not equality (i.e. if the same string is being checked).

True, then just needs for the identity one.

File: addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java (hunk 5 above)
Comment (thc202) [reply addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java:150]: The source should return two arguments (instead of an array).

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java (hunk 3 above)
Comment (thc202) [reply addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java:58]: Although equals will already do that.... guess it's just left the comment.

File: addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java (hunk 5 above)
Comment (kingthorin) [reply addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java:150]: Okay I'll try to sort that out :wink:

Comment (thc202) [reply addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java:151]: Don't create the arrays, just pass the strings.

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java (hunk 3 above)
Comment (thc202) [reply addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/DiceMatcher.java:57]: This is no longer a quick check (not as the identity).

File: addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java (hunk 5 above)
Comment (kingthorin) [reply addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java:151]: But how to pass two sets of strings, or generate the wanted (AB, BA, AA) combinations... I'm sure there's a simple solution. I just gotta research a bit.

Comment (thc202) [reply addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java:151]: Do `Arguments.of(null, ORIGINAL_STRING}), Arguments.of(...)` (remove the arrays).

Comment (kingthorin) [reply addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java:151]: Thanks, I knew it had to be something simple 😀

Comment (thc202) [reply addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java]: The comment about the test method name was not addressed (I'm fine with that name if that's really just testing a short string and not a completely different one).

Comment (kingthorin) [reply addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java]: I had named it like that specifically, because I intended that test method do be based around the initial few checks (null, short strings, etc)

However, I guess there's no reason that it couldn't include other zero percent options.

Comment (thc202) [reply addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java]: OK, that sounds good to me, no changes needed.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

File: addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java [hunk 6]
```diff
@@ -132,4 +136,22 @@ void shouldGiveCorrectPercentageForDifferentString() {
...
         assertThat(sim, is(equalTo(25)));
     }
+
+    static Stream<Arguments> stringSetSource() {
+        return Stream.of(
+                Arguments.of(null, ORIGINAL_STRING),
+                Arguments.of(ORIGINAL_STRING, null),
+                Arguments.of(null, null),
+                Arguments.of("ab", ORIGINAL_STRING),
+                Arguments.of(ORIGINAL_STRING, "ab"));
```
Comment (ricekot) [inline addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java:146]: "ab" -> "a"?
We are only checking if a parameter is a single character ([L59](https://github.com/zaproxy/zap-extensions/pull/2907/files#diff-1ff4022fe74aa6c1d61fc6c935f1c76ef91fb39849f300b9d930698496b4c35eR59)). This test case shouldn't have passed...

Comment (kingthorin) [reply addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java:146]: It probably still calculates to zero, but you're right. For the intent it should be a single char.

Comment (kingthorin) [reply addOns/commonlib/src/test/java/org/zaproxy/addon/commonlib/DiceMatcherUnitTest.java]: Fix pushed

**Review by ricekot — APPROVED** [review]
